src/Microsoft.Health.Fhir.TemplateManagement/Client/OrasClient.cs

```diff
@@ -17,46 +17,65 @@ namespace Microsoft.Health.Fhir.TemplateManagement.Client
 
         public OrasClient(string imageReference)
         {
             EnsureArg.IsNotNull(imageReference, nameof(imageReference));
 
             _imageReference = imageReference;
+            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("ORAS_CACHE")))
+            {
+                Environment.SetEnvironmentVariable("ORAS_CACHE", ".oras/cache");
+            }
         }
 
         public async Task PullImageAsync(string outputFolder)
         {
             string command = $"pull  \"{_imageReference}\" -o \"{outputFolder}\"";
-            await OrasExecutionAsync(command, Directory.GetCurrentDirectory());
+            string output = await OrasExecutionAsync(command, Directory.GetCurrentDirectory());
+            Console.WriteLine(output);
+            string digest;
+            try
+            {
+                digest = _digestRx.Matches(output)[0].Groups["digest"].ToString();
+            }
+            catch (Exception ex)
+            {
+                throw new OrasException(TemplateManagementErrorCode.OrasProcessFailed, "Return image digest failed.", ex);
+            }
+
+            File.WriteAllText(Path.Combine(outputFolder, Constants.ManifestFileName), LoadManifestContentFromCache(digest));
         }
 
-        public async Task PushImageAsync(string inputFolder, List<string> filePathList)
+        public async Task PushImageAsync(string inputFolder)
         {
+            if (!Directory.Exists(inputFolder))
+            {
+                throw new DirectoryNotFoundException("Image Folder not found.");
+            }
+
             string argument = string.Empty;
             string command = $"push \"{_imageReference}\"";
 
-            // In order to remove image's directory prefix. (e.g. "layers/layer1.tar.gz" --> "layer1.tar.gz"
+            var filePathToPush = Directory.EnumerateFiles(inputFolder, "*.tar.gz", SearchOption.AllDirectories);
+
+            // In order to remove image's directory prefix. (e.g. "layers/layer1" --> "layer1"
             // Change oras working folder to inputFolder
-            foreach (var filePath in filePathList)
+            foreach (var filePath in filePathToPush)
             {
-                if (!File.Exists(Path.Combine(inputFolder, filePath)))
-                {
-                    throw new OverlayException(TemplateManagementErrorCode.ImageLayersNotFound, "Image layer not found");
-                }
-
-                argument += $" \"{filePath}\"";
+                argument += $" \"{Path.GetRelativePath(inputFolder, filePath)}\"";
             }
 
             if (string.IsNullOrEmpty(argument))
             {
                 throw new OverlayException(TemplateManagementErrorCode.ImageLayersNotFound, "No file for push.");
             }
 
-            await OrasExecutionAsync(string.Concat(command, argument), inputFolder);
+            Console.WriteLine(await OrasExecutionAsync(string.Concat(command, argument), inputFolder));
```

Comment:
Do we intend to write message to console directly?

Reply:
Thanks, fixed it.